src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,916 @@
+/**
+ * @file feature_extraction_Impl.hpp
+ * @author Nilay Jain
+ *
+ * Implementation of feature extraction methods.
+ */
+#ifndef MLPACK_METHODS_EDGE_BOXES_STRUCTURED_TREE_IMPL_HPP
+#define MLPACK_METHODS_EDGE_BOXES_STRUCTURED_TREE_IMPL_HPP
+
+
+#include "feature_extraction.hpp"
+#include <mlpack/methods/pca/pca.hpp>
+
+namespace mlpack {
+namespace structured_tree {
+
+
+template<typename MatType, typename CubeType>
+StructuredForests<MatType, CubeType>::
+StructuredForests(mlpack::structured_tree::FeatureParameters F) : params(F) {}
+
+
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+GetFeatureDimension(arma::vec& FtrDim)
+{
+  FtrDim = arma::vec(2);
+
+  const size_t shrink = this->params.Shrink();
+  const size_t pSize = this->params.PSize();
+  const size_t numCell = this->params.NumCell();
+  const size_t numOrient = this->params.NumOrient();
+  const size_t nColorCh = params.RGBD() == 0 ? 3 : 4;
+  const size_t nCh = nColorCh + 2 * (1 + numOrient);
+  FtrDim[0] = std::pow((pSize / shrink) , 2) * nCh;
+  FtrDim[1] = std::pow(numCell , 2) * (std::pow (numCell, 2) - 1) / 2 * nCh;
+}
+
+
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+DistanceTransform1D(const arma::vec& f, const size_t n, const double inf,
+                    arma::vec& d)
+{
+  arma::vec v(n), z(n + 1);
+  d = arma::vec(n);
+  size_t k = 0;
+  v[0] = 0.0;
+  z[0] = -inf;
+  z[1] = inf;
+  for (size_t q = 1; q <= n - 1; ++q)
+  {
+    float s  = ( (f[q] + q * q)-( f[v[k]] + v[k] * v[k]) ) / (2 * q - 2 * v[k]);
+    while (s <= z[k])
+    {
+      --k;
+      s  = ( (f[q] + q * q) - (f[v[k]] + v[k] * v[k]) ) / (2 * q - 2 * v[k]);
+    }
+
+    ++k;
+    v[k] = static_cast<double>(q);
+    z[k] = s;
+    z[k+1] = inf;
+  }
+
+  k = 0;
+  for (size_t q = 0; q <= n-1; q++)
+  {
+    while (z[k+1] < q)
+      ++k;
+    d[q] = (q - v[k]) * (q - v[k]) + f[v[k]];
+  }
+}
+
+
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+DistanceTransform2D(MatType &Im, const double inf)
+{
+  arma::vec f(std::max(Im.n_rows, Im.n_cols));
+  // transform along columns
+  for (size_t x = 0; x < Im.n_cols; ++x)
+  {
+    f.subvec(0, Im.n_rows - 1) = Im.col(x);
+    arma::vec d;
+    this->DistanceTransform1D(f, Im.n_rows, inf, d);
+    Im.col(x) = d;
+  }
+
+  // transform along rows
+  for (size_t y = 0; y < Im.n_rows; y++)
+  {
+    f.subvec(0, Im.n_cols - 1) = Im.row(y).t();
+    arma::vec d;
+    this->DistanceTransform1D(f, Im.n_cols, inf, d);
+    Im.row(y) = d.t();
```

Comment:
It looks like you could pass im.row(y) instead of passing a copy of the transposed row. In this case you have to adjust the DistanceTransform1D function header:

```
 arma::vec DistanceTransform1D(arma::row& f, const size_t n, ...) 
```